datatypes/src/commonMain/kotlin/at/asitplus/crypto/datatypes/CryptoSignature.kt

```diff
@@ -6,54 +6,59 @@ import at.asitplus.crypto.datatypes.asn1.DERTags.DER_SEQUENCE
 import at.asitplus.crypto.datatypes.io.Base64Strict
 import at.asitplus.crypto.datatypes.misc.BitLength
 import at.asitplus.crypto.datatypes.misc.max
+import at.asitplus.crypto.datatypes.pki.X509Certificate
 import com.ionspin.kotlin.bignum.integer.BigInteger
 import com.ionspin.kotlin.bignum.integer.Sign
 import io.matthewnelson.encoding.core.Decoder.Companion.decodeToByteArray
 import io.matthewnelson.encoding.core.Encoder.Companion.encodeToString
-import kotlinx.serialization.Contextual
 import kotlinx.serialization.KSerializer
 import kotlinx.serialization.Serializable
 import kotlinx.serialization.descriptors.PrimitiveKind
 import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
 import kotlinx.serialization.descriptors.SerialDescriptor
 import kotlinx.serialization.encoding.Decoder
 import kotlinx.serialization.encoding.Encoder
-import kotlin.math.max
 
 
 /**
- * Data class which holds Asn1 Encoding of a signature of a specified algorithm
+ * Interface which holds Asn1 Encoding of a signature of a specified algorithm
  * Allows simple ASN1 - Raw transformation of signature values
- * Does not check for anything!
  */
 
 @Serializable(with = CryptoSignature.CryptoSignatureSerializer::class)
-sealed class CryptoSignature(
-    @Contextual
-    protected val signature: Asn1Element,
-) : Asn1Encodable<Asn1Element> {
+sealed interface CryptoSignature : Asn1Encodable<Asn1Element> {
+    val signature: Asn1Element
+
 
     /**
-     * Removes ASN1 Structure and returns the value(s) as ByteArray
+     * Allows classifying CryptoSignatures into two classes
      */
-    abstract val rawByteArray: ByteArray
-
-    abstract fun encodeToTlvBitString(): Asn1Element
-
-    override fun equals(other: Any?): Boolean {
-        if (this === other) return true
-        if (other == null || this::class != other::class) return false
-
-        other as CryptoSignature
+    sealed interface Defined : CryptoSignature {
```

Comment:
what's the benefit of this inheritance structure? why not just `IllDefined: CryptoSignature` and `WellDefined: CryptoSignature`? (Or even drop `IllDefined` altogether since it seems to add nothing?)

Reply:
Without two mutually exlcusive Types, you won't win anything, so it needs to stay, because otherwise

this
```
//it is CryptoSignature.Defined
when (it) {
    is CryptoSignature.Defined.Well -> it
    is CryptoSignature.Defined.Ill -> when (it) {
        is CryptoSignature.EC.IndefiniteLength -> it.withCurve((publicKey as CryptoPublicKey.EC).curve)
    }
}
```

turns into this useless mess:
```
//it is CryptoSignature
when (it) {
    is CryptoSignature.Definite -> it
    else -> when (it) {
        is CryptoSignature.EC.IndefiniteLength -> it.withCurve((publicKey as CryptoPublicKey.EC).curve)
        is CryptoSignature.EC.DefiniteLength -> TODO("utterly useless")
        is CryptoSignature.RSAorHMAC -> TODO("utterly useless")
    }
}
```

Reply:
```kotlin
val defined = when (it) {
        is CryptoSignature.WellDefined -> it
        is CryptoSignature.EC.IndefiniteLength -> it.withCurve()
    }
```
https://pl.kotl.in/MyyqCQvUs

Reply:
to me, this is not a technical issue, but one of how expressive the code is to a human. without the inverse interface, something the grouping (first code block I posted) won't be possible: https://pl.kotl.in/AX88suTB4